### PR TITLE
provisioner: Add support to fetch kbs nodeport service ip

### PR DESCRIPTION
**Pull request by kartikjoshi21**
Fixes: #1809

### CODE REVIEW

**Comment by kartikjoshi21** [comment]
Steps to test this PR:

```
export AZURE_RESOURCE_GROUP="test-kartikPR-1812"
export AZURE_REGION="eastus"
az group create --name "${AZURE_RESOURCE_GROUP}" \
    --location "${AZURE_REGION}"

export AZURE_SUBSCRIPTION_ID=$(az account show --query id --output tsv)
export USER_ASSIGNED_IDENTITY_NAME="caa-${AZURE_RESOURCE_GROUP}"
az identity create \
    --name "${USER_ASSIGNED_IDENTITY_NAME}" \
    --resource-group "${AZURE_RESOURCE_GROUP}" \
    --location "${AZURE_REGION}" \
    --subscription "${AZURE_SUBSCRIPTION_ID}"

export PRINCIPAL_ID="$(az identity show \
    --name "${USER_ASSIGNED_IDENTITY_NAME}" \
    --resource-group "${AZURE_RESOURCE_GROUP}" \
    --subscription "${AZURE_SUBSCRIPTION_ID}" --query principalId -otsv)"

sleep 30
az role assignment create \
    --role Contributor \
    --assignee-object-id "${PRINCIPAL_ID}" \
    --scope "/subscriptions/${AZURE_SUBSCRIPTION_ID}"

export AZURE_CLIENT_ID="$(az identity show \
    --resource-group "${AZURE_RESOURCE_GROUP}" \
    --name "${USER_ASSIGNED_IDENTITY_NAME}" --query 'clientId' -otsv)"

export CLUSTER_NAME="e2e"
export AZURE_IMAGE_ID="/CommunityGalleries/cocopodvm-d0e4f35f-5530-4b9c-8596-112487cdea85/Images/podvm_image0/Versions/2024.04.17"

# Docker image for KBS
# https://github.com/confidential-containers/kbs/pkgs/container/staged-images%2Fkbs

cat <<EOF >/tmp/provision_azure.properties
AZURE_CLIENT_ID="${AZURE_CLIENT_ID}"
AZURE_SUBSCRIPTION_ID="${AZURE_SUBSCRIPTION_ID}"
RESOURCE_GROUP_NAME="${AZURE_RESOURCE_GROUP}"
CLUSTER_NAME="${CLUSTER_NAME}"
LOCATION="${AZURE_REGION}"
SSH_KEY_ID="id_rsa.pub"
AZURE_IMAGE_ID="${AZURE_IMAGE_ID}"

AZURE_CLI_AUTH="true"
MANAGED_IDENTITY_NAME="${USER_ASSIGNED_IDENTITY_NAME}"

# Deploy the same one that is merged on the CAA main
KBS_IMAGE="ghcr.io/confidential-containers/staged-images/kbs"
KBS_IMAGE_TAG="dc01f454264fb4350e5f69eba05683a9a1882c41"

# Get the tag from: https://quay.io/repository/confidential-containers/cloud-api-adaptor?tab=tags&tag=latest
CAA_IMAGE="quay.io/confidential-containers/cloud-api-adaptor:d32c2e6f92dd610a688c78d92c463f39781b8416"
EOF

pushd src/cloud-api-adaptor/
ssh-keygen -t rsa -b 4096 -f install/overlays/azure/id_rsa -N "" -C dev@coco.io

pushd test/tools
git clone git@github.com:confidential-containers/trustee.git

pushd trustee
git checkout dc01f454264fb4350e5f69eba05683a9a1882c41
popd
popd

# Now open a new terminal

export TEST_PROVISION_FILE=/tmp/provision_azure.properties
export CLOUD_PROVIDER=azure
export BUILTIN_CLOUD_PROVIDERS=azure
export DEPLOY_KBS=true
export TEST_PROVISION=true

pushd test/tools
make caa-provisioner-cli
./caa-provisioner-cli -action=provision

popd

```

```
# k get svc -n coco-tenant
NAME   TYPE       CLUSTER-IP    EXTERNAL-IP   PORT(S)          AGE
kbs    NodePort   10.0.95.196   <none>        8080:30146/TCP   13m
```

File: src/cloud-api-adaptor/test/tools/provisioner-cli/main.go [hunk 1]
```diff
@@ -98,13 +98,13 @@ func main() {
 			log.Fatal(err)
 		}
 
-		var kbsPodIP string
-		kbsPodIP, err = keyBrokerService.GetKbsPodIP(context.TODO(), cfg)
+		var kbsSvcIP string
+		kbsSvcIP, err = keyBrokerService.GetKbsServiceIP(context.TODO(), cfg)
```
Comment (mkulke) [inline src/cloud-api-adaptor/test/tools/provisioner-cli/main.go:102]: kbsEndpoint

**Review by mkulke — APPROVED** [review]

File: src/cloud-api-adaptor/test/provisioner/provision.go [hunk 2]
```diff
@@ -287,46 +287,46 @@ func (lio *KbsInstallOverlay) Edit(ctx context.Context, cfg *envconf.Config, pro
...
 		}
 	}
 
-	if matchingPod == nil {
-		return "", fmt.Errorf("No pod with label selector found")
-	}
-
-	fmt.Printf("Pod IP: %s\n", matchingPod.Status.PodIP)
-	return matchingPod.Status.PodIP, nil
+	serviceString := fmt.Sprintf("%s:%d", clusterIP, nodePort)
```
Comment (huoqifeng) [inline src/cloud-api-adaptor/test/provisioner/provision.go:328]: I think this is still the address in CNI network but not NodeIP:port?
For example, in my libvirt env.
```
# kubectl get svc kbs
NAME   TYPE       CLUSTER-IP     EXTERNAL-IP   PORT(S)          AGE
kbs    NodePort   10.96.46.215   <none>        8080:30194/TCP   3d2h

# kcli list vm
+------------------------+--------+-----------------+------------+-----------+---------+
|          Name          | Status |        Ip       |   Source   |    Plan   | Profile |
+------------------------+--------+-----------------+------------+-----------+---------+
|  peer-pods-ctlplane-0  |   up   | 192.168.122.123 | ubuntu2204 | peer-pods |  kvirt  |
|   peer-pods-worker-0   |   up   | 192.168.122.108 | ubuntu2204 | peer-pods |  kvirt  |
```
I think we should have `http://192.168.122.108:30194` but not `http://10.96.46.215:8080`.

**Review by huoqifeng — COMMENTED** [review]
Thanks @kartikjoshi21 I added a comment...

**Comment by kartikjoshi21** [comment]
@mkulke @huoqifeng updated code and logs, Thanks.

File: src/cloud-api-adaptor/test/provisioner/provision.go [hunk 3]
```diff
@@ -287,46 +287,77 @@ func (lio *KbsInstallOverlay) Edit(ctx context.Context, cfg *envconf.Config, pro
 	return nil
 }
 
-func (p *KeyBrokerService) GetKbsPodIP(ctx context.Context, cfg *envconf.Config) (string, error) {
+func getNodeIPForSvc(service corev1.Service, cfg *envconf.Config) (string, error) {
 	client, err := cfg.NewClient()
 	if err != nil {
 		return "", err
 	}
+	podList := &corev1.PodList{}
```
Comment (huoqifeng) [inline src/cloud-api-adaptor/test/provisioner/provision.go:295]: Shall we filter `KBS` pod when list Pod here?

File: src/cloud-api-adaptor/test/provisioner/provision.go [hunk 4]
```diff
@@ -287,46 +287,77 @@ func (lio *KbsInstallOverlay) Edit(ctx context.Context, cfg *envconf.Config, pro
...
-	namespace := "coco-tenant"
-	deploymentName := "kbs"
+	nodeList := &corev1.NodeList{}
+	if err := client.Resources("").List(context.TODO(), nodeList); err != nil {
+		return "", err
+	}
 
-	err = AllPodsRunning(ctx, cfg, namespace)
+	for _, pod := range podList.Items {
+		if pod.Spec.NodeName != "" {
```
Comment (huoqifeng) [inline src/cloud-api-adaptor/test/provisioner/provision.go:306]: or shall we find the KBS pod first here but not use the 1st pod?

Comment (kartikjoshi21) [reply src/cloud-api-adaptor/test/provisioner/provision.go]: changed to filter pod first on deployment name.

**Comment by huoqifeng** [comment]
Looks some CI errors?

**Comment by surajssd** [comment]
Functionality wise, this has worked for me.

**Comment by kartikjoshi21** [comment]
> Looks some CI errors?

It was unrelated, retriggered.

**Review by huoqifeng — APPROVED** [review]
LGTM, thanks @kartikjoshi21

**Pull request merged**